feat: opt-in sync of deletes and restores from web to Android (beta timeline)

PeterOmbodi opened the pull request:
> [!NOTE]
>  **UPD**: Sync only applies to files from albums included in the device’s backup list. To enable it, tap the button with the cloud icon in the app bar, then select “Select” and choose the folder to back up. I used Immich.

**Description**

Functionality for synchronizing the local media store when remotely deleting to trash and restoring files has been added. (Android, beta)
This solves the issue of file deletion synchronization on web and mobile apps.

Fixes # (issue)

**How Has This Been Tested?**

To verify this, you need an Android 11+ device.

1. Launch the Flutter app on an Android device.
2. Log in to the app (I used the demo server and demo account).
3. Go to Settings => ~~Switch timeline to beta mode~~, then -Advanced and enable the 'Sync remote deletions' switch.
4. Download any photo or multiple photos from the photo list.
5. Grant access to local albums, specifically to the Immich album (Cloud icon to the left of the profile icon).
6. Open the Immich web app and log in with the same account (I used the demo).
7. In the web app, delete (move to trash) a photo that is locally downloaded on your phone (marked with a cloud + icon).
8. For devices using Google Photos as the gallery app, make sure that syncing is disabled or paused — otherwise, a file moved to the trash may appear both in the Trash and in the Photos section.
9. Open the gallery, go to the trash, and you should see the file that is in the trash on the web app.
10. Return to the mobile app.
11. In the web app, restore the file from the trash, and make sure it appears in the Flutter app's photo list (with the cloud + icon).
12. Open the gallery, and the file should be in the Immich folder (restored).

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

Commit:
feature(mobile, beta, Android): handle remote asset trash/restore events and rescan media
- Handle move to trash and restore from trash for remote assets on Android
- Trigger MediaScannerConnection to rescan affected media files

Commit:
feature(mobile, beta, Android): fix rescan

Commit:
fix imports

Commit:
fix checking conditions

Commit:
refactor naming

Commit:
fix line breaks

PeterOmbodi commented on mobile/android/app/src/main/kotlin/app/alextran/immich/BackgroundServicePlugin.kt:
```diff
@@ -309,6 +311,35 @@ class BackgroundServicePlugin : FlutterPlugin, MethodChannel.MethodCallHandler,
+                val path = it.getString(it.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA))
+                if (!path.isNullOrBlank()) {
+                  Log.i(TAG, "Scanning updated file: $path")
+                  MediaScannerConnection.scanFile(context, arrayOf(path), null, null)
```
In my case, I experienced a slight delay with the gallery update, which is why I added this code. It might be related to my dev environment, and if everything works fine in production, I’ll remove it.

Commit:
refactor code
rollback changes in BackgroundServicePlugin

shenlong-tanwen commented on mobile/lib/domain/services/asset.service.dart:
```diff
@@ -84,4 +100,40 @@ class AssetService {
     return _localAssetRepository.getHashedCount();
   }
+
+  Future<void> handleRemoteTrashChanges(Iterable<({String checksum, DateTime? deletedAt})> syncItems) async {
+    if (_platform.isAndroid && _appSettingsService.getSetting<bool>(AppSettingsEnum.manageLocalMediaAndroid)) {
+      final trashedItems = syncItems.where((item) => item.deletedAt != null);
+
+      if (trashedItems.isNotEmpty) {
+        final trashedAssetsChecksums = trashedItems.map((syncItem) => syncItem.checksum);
```
Yep, I'd prefer that TBH. Can you refactor this one to use a separate `TrashService`?

Commit:
refactor code (use separate TrashService)

shenlong-tanwen commented on mobile/lib/domain/services/trash.service.dart:
```diff
@@ -0,0 +1,65 @@
+  final LocalFilesManagerRepository _localFilesManager;
+  final StorageRepository _storageRepository;
+  final Platform _platform;
+  final Logger _logger;
```
We usually initialise the logger directly inside the service class. The Trash service is never going to be created a constant because of the other dependencies anyways

shenlong-tanwen commented on mobile/lib/domain/services/trash.service.dart:
```diff
@@ -0,0 +1,65 @@
+       _logger = logger;
+
+  Future<void> handleRemoteChanges(Iterable<({String checksum, DateTime? deletedAt})> syncItems) async {
+    if (_platform.isAndroid && _appSettingsService.getSetting<bool>(AppSettingsEnum.manageLocalMediaAndroid)) {
```
Check the platform and the setting and do a early return instead of the if .. else handling

shenlong-tanwen commented on mobile/lib/domain/services/trash.service.dart:
```diff
@@ -0,0 +1,65 @@
+    if (_platform.isAndroid && _appSettingsService.getSetting<bool>(AppSettingsEnum.manageLocalMediaAndroid)) {
+      final trashedItems = syncItems.where((item) => item.deletedAt != null);
+
+      if (trashedItems.isNotEmpty) {
+        final trashedAssetsChecksums = trashedItems.map((syncItem) => syncItem.checksum);
+        final localAssetsToTrash = await _localAssetRepository.getByChecksums(trashedAssetsChecksums);
+        if (localAssetsToTrash.isNotEmpty) {
+          final mediaUrls = await Future.wait(
+            localAssetsToTrash.map(
+              (localAsset) => _storageRepository.getAssetEntityForAsset(localAsset).then((e) => e?.getMediaUrl()),
+            ),
+          );
+          _logger.info("Moving to trash ${mediaUrls.join(", ")} assets");
+          await _localFilesManager.moveToTrash(mediaUrls.nonNulls.toList());
+        }
+      }
+      final modifiedItems = syncItems.where((e) => e.deletedAt == null);
+      if (modifiedItems.isNotEmpty) {
+        final modifiedChecksums = modifiedItems.map((syncItem) => syncItem.checksum);
+        final remoteAssetsToRestore = await _remoteAssetRepository.getByChecksums(modifiedChecksums, isTrashed: true);
+        if (remoteAssetsToRestore.isNotEmpty) {
+          _logger.info("Restoring from trash ${remoteAssetsToRestore.map((e) => e.name).join(", ")} assets");
+          for (final remoteAsset in remoteAssetsToRestore) {
+            await _localFilesManager.restoreFromTrash(remoteAsset.name, remoteAsset.type.index);
+          }
+        }
+      }
```
Can you extract it into two different methods? One to trash and the other to restore. We could hopefully be able to reuse the same trash method for the review screen as well

shenlong-tanwen commented on mobile/lib/domain/services/trash.service.dart:
```diff
@@ -0,0 +1,65 @@
+        if (remoteAssetsToRestore.isNotEmpty) {
+          _logger.info("Restoring from trash ${remoteAssetsToRestore.map((e) => e.name).join(", ")} assets");
+          for (final remoteAsset in remoteAssetsToRestore) {
+            await _localFilesManager.restoreFromTrash(remoteAsset.name, remoteAsset.type.index);
```
The Sync object already has the name and the index. Can we pass it over to his method instead of the record so we don't have to do this DB call here?

shenlong-tanwen replied:
Hmm, That is true! The list includes all the assets that are modified and not just those that were restored. This is fine then!

shenlong-tanwen commented on mobile/lib/domain/services/trash.service.dart:
```diff
@@ -0,0 +1,65 @@
+        if (localAssetsToTrash.isNotEmpty) {
+          final mediaUrls = await Future.wait(
+            localAssetsToTrash.map(
+              (localAsset) => _storageRepository.getAssetEntityForAsset(localAsset).then((e) => e?.getMediaUrl()),
```
I wonder if we should just pass the list of local Ids to the native side and fetch the media URLs with a single call to the `MediaStore`

PeterOmbodi replied:
The method `MediaStore.createTrashRequest(...)` requires a list of Uri objects that refer to media files in the MediaStore.
If we pass a list of _id values to the native part of the code, we’ll also need to include the media type (Images, Video, Audio, Files, etc.) and construct the corresponding Uri using `ContentUris.withAppendedId(...)`.

shenlong-tanwen replied:
>  we’ll also need to include the media type (Images, Video, Audio, Files, etc.)

We can use the list of Ids to fetch the type from MediaStore and construct the Media URL, and trash the assets, all in a single call to the native side.

The current code goes through the list of assets, and for each of them, uses two future to fetch the media url, and then once all the futures from the list are resolved, we make one more native call to trash the assets

Commit:
refactor code

shenlong-tanwen commented on mobile/lib/domain/services/trash.service.dart:
```diff
@@ -0,0 +1,74 @@
+      );
+      if (remoteAssetsToRestore.isNotEmpty) {
+        _logger.info("Restoring from trash ${remoteAssetsToRestore.map((e) => e.name).join(", ")} assets");
+        for (final remoteAsset in remoteAssetsToRestore) {
+          await _localFilesManager.restoreFromTrash(remoteAsset.name, remoteAsset.type.index);
+        }
+      }
```
Map the restores to a list of futures and do a `Future.wait` on them

shenlong-tanwen reviewed (commented):
Thanks a lot for working on this. The changes looks great, only few pending comments to be resolved and we can get this merged.

BTW, Can you also fix the failing lint? You've to format the file before pushing them

Commit:
parallelize restoreFromTrash calls with Future.wait
format trash.provider.dart

Commit:
try to re-format trash.provider.dart

Commit:
re-format trash.provider.dart

Commit:
rename TrashService to TrashSyncService to avoid duplicated names
revert changes in original trash.provider.dart

PeterOmbodi commented:
Hi @shenlong-tanwen, I’ve pushed the fixes for the pending comments and lint errors.
When you have some time, could you please continue the review?

shenlong-tanwen commented on mobile/lib/domain/services/trash_sync.service.dart:
```diff
@@ -0,0 +1,74 @@
+    if (!_platform.isAndroid || !_appSettingsService.getSetting<bool>(AppSettingsEnum.manageLocalMediaAndroid)) {
+      return Future.value();
+    }
+    final trashedAssetsChecksums = syncItems
+        .where((item) => item.deletedAt != null)
+        .map((syncItem) => syncItem.checksum);
+    await applyRemoteTrashToLocal(trashedAssetsChecksums);
+    final modifiedAssetsChecksums = syncItems
+        .where((item) => item.deletedAt == null)
+        .map((syncItem) => syncItem.checksum);
+    await applyRemoteRestoreToLocal(modifiedAssetsChecksums);
```
NIT: We could just loop through the list and construct both the list in a single run

shenlong-tanwen commented on mobile/lib/domain/services/trash_sync.service.dart:
```diff
@@ -0,0 +1,74 @@
+       _storageRepository = storageRepository,
+       _platform = const LocalPlatform();
+
+  Future<void> handleRemoteChanges(Iterable<({String checksum, DateTime? deletedAt})> syncItems) async {
```
I'm not a big fan of anonymous records. Can we have a typedef for this? Something like `TrashSyncItem`?

shenlong-tanwen commented on mobile/lib/domain/services/trash_sync.service.dart:
```diff
@@ -0,0 +1,74 @@
+    await applyRemoteRestoreToLocal(modifiedAssetsChecksums);
+  }
+
+  Future<void> applyRemoteTrashToLocal(Iterable<String> trashedAssetsChecksums) async {
```
This and the following methods could be private methods

shenlong-tanwen commented on mobile/lib/domain/services/trash_sync.service.dart:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  Future<void> applyRemoteTrashToLocal(Iterable<String> trashedAssetsChecksums) async {
+    if (trashedAssetsChecksums.isNotEmpty) {
+      final localAssetsToTrash = await _localAssetRepository.getByChecksums(trashedAssetsChecksums);
+      if (localAssetsToTrash.isNotEmpty) {
```
Prefer early returns

shenlong-tanwen commented on mobile/lib/domain/services/trash_sync.service.dart:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  Future<void> applyRemoteRestoreToLocal(Iterable<String> modifiedAssetsChecksums) async {
+    if (modifiedAssetsChecksums.isNotEmpty) {
+      final remoteAssetsToRestore = await _remoteAssetRepository.getByChecksums(
+        modifiedAssetsChecksums,
+        isTrashed: true,
+      );
+      if (remoteAssetsToRestore.isNotEmpty) {
+        _logger.info("Restoring from trash ${remoteAssetsToRestore.map((e) => e.name).join(", ")} assets");
+        await Future.wait(
+          remoteAssetsToRestore.map((asset) => _localFilesManager.restoreFromTrash(asset.name, asset.type.index)),
+        );
```
Prefer early returns

shenlong-tanwen reviewed (commented):
I'm happy with how this looks. Just some minor nitpicks. I'll get this tested today or tomorrow and we can get this merged! Thanks a lot for working on this and apologies for not getting back sooner

shenlong-tanwen commented:
@PeterOmbodi I did some testing and faced the following issue during my testing:

- When I restore a mix of merged and remote only assets after trashing them, the local part of the merged assets are properly restored, however, the sync gets stuck and never completes. Can you check this?

PeterOmbodi commented:
> @PeterOmbodi I did some testing and faced the following issue during my testing:
> 
> * When I restore a mix of merged and remote only assets after trashing them, the local part of the merged assets are properly restored, however, the sync gets stuck and never completes. Can you check this?

I’ll pull the latest changes from main and check — hopefully that fixes it.

Commit:
refactor code (minor nitpicks)

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

alextran1502 committed:
Merge branch 'main' into feature/sync_assets_trashed_state

alextran1502 commented:
Hi @PeterOmbodi, thank you for bringing the PR to the finished line. I just tested the PR I notice one possible bug. You can recreate follow the steps below

- Be on the beta timeline
- On web, move asset to trash
- Open the mobile app, make sure that they are synced, and the asset is in the trash. Check the local gallery to make sure they are in the system's trash as well
- On web, restore the asset
- Open the mobile app from the background, and you will see the sync indicator never finishes. You will have to force-kill the app and reopen it to see the restored assets

Commit:
process restoreFromTrash sequentially instead of Future.wait

PeterOmbodi commented:
Hi @alextran1502, I was able to reproduce the infinite sync when I discovered a restore-from-trash error. When switching between our app and another one, at some point the sync starts and the indicator may stay on. The logs show restore errors, and when combined with BackgroundWorkerBgService logs, the indicator never turns off.
It seems related to restoring files in parallel - after disabling parallelism, the indicator started turning off correctly. At the moment, I’ve pushed changes for sequential restore calls. Could you please check if synchronization works properly now?

Now regarding the restore issues themselves. If we have duplicate assets in the timeline and we keep their local copies, then the duplicate file name in local storage gets a sequence number added. During restore, however, the “clean” cloud file name is used, since the DB table entry for local assets is deleted along with the asset. As a result, we can only restore the first instance of the local asset, but not its duplicates.

Maybe instead of deleting the record from the local_asset_entity table when an asset is moved to trash, we should keep it and add a deletedAt timestamp, or introduce a separate table for deleted assets? Or maybe you could suggest another solution?

PeterOmbodi commented:
Hi @alextran1502 Also, storing information about locally deleted assets would allow us to optimize the restore process. Right now the restore logic works like this: we get the list of modified assets, and before updating the local DB with info about a modified remote asset, we check if it had the `deletedAt` flag, and if so - we try to restore it.

The problem is, at this point we cannot know whether the user ever had a local copy saved, since the **local_asset_entity** table only keeps information about local assets that have not been moved to the trash. As a result, we may end up trying to restore files that were never on the device. This leads both to errors like _"[LocalFilesManager] Error restore file from trash"_ and to suboptimal code execution.

Having information about locally deleted assets would let us attempt to restore only those files, and also solve the issue of restoring deleted duplicates - which might be even more important.

shenlong-tanwen commented:
> @shenlong-tanwen When moving a file to the trash everything is straightforward - we have the info in **local_asset_entity**. The problem may occur during restore: at that point we cannot (or at least I don’t yet know how to) determine which specific duplicate file needs to be restored. Right now we restore by file name and type, but ideally it should be by checksum.

The actual mismatch is how the server ignores duplicates but how the app still trashes them. i.e, the `getByChecksums` method on the local asset repository would return multiple local assets for a single remote asset if they are duplicates. If we only trash the same asset that got trashed on the server, there should not be a case where there are multiple duplicate files in the system trash.

> I see that **remote_asset_entity** contains the fields `createdAt `and `modifiedAt`. I’ll try to match these with MediaStore.MediaColumns.DATE_MODIFIED or DATE_ADDED and let you know the result.

This just switches the way we handle this. My suggestion would only put one file in trash. With your suggestion, we put everything in the trash but only restore the single asset that got restored on the server. If we trash everything, shouldn't we also be restoring everything as well?

Commit:
group local assets by checksum before moving to trash
delete LocalAssetEntity records when moved to trash
refactor code

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
Merge remote-tracking branch 'public/main' into feature/sync_assets_trashed_state

Commit:
fix format

Commit:
Merge remote-tracking branch 'public/main' into feature/sync_assets_trashed_state

Commit:
use checksum for asset restoration
refactro code

Commit:
fix format

Commit:
sync trash only for backup-selected assets

Commit:
Merge remote-tracking branch 'public/main' into feature/sync_assets_trashed_state

Commit:
Merge remote-tracking branch 'public/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/test/domain/services/sync_stream_service_test.dart

Commit:
feat(db): add local_trashed_asset table and integrate with restoration flow
- Add new `local_trashed_asset` table to store metadata of trashed assets
- Save trashed asset info into `local_trashed_asset` before deletion
- Use `local_trashed_asset` as source for asset restoration
- Implement file restoration by `mediaId`

Commit:
Merge remote-tracking branch 'public/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/drift_schemas/main/drift_schema_v10.json
#	mobile/lib/infrastructure/repositories/db.repository.dart
#	mobile/lib/infrastructure/repositories/db.repository.drift.dart
#	mobile/lib/infrastructure/repositories/db.repository.steps.dart
#	mobile/test/drift/main/generated/schema_v10.dart

Commit:
resolve merge conflicts

Commit:
fix index creating on migration

Commit:
Merge remote-tracking branch 'public/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/lib/infrastructure/repositories/local_asset.repository.dart

Commit:
rework trashed assets handling
- add new table trashed_local_asset
- mirror trashed assets data in trashed_local_asset.
- compute checksums for assets trashed out-of-app.
- restore assets present in trashed_local_asset and non-trashed in remote_asset.
- simplify moving-to-trash logic based on remote_asset events.

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/android/app/src/main/kotlin/app/alextran/immich/sync/Messages.g.kt
#	mobile/ios/Runner/Sync/Messages.g.swift
#	mobile/lib/domain/services/hash.service.dart
#	mobile/lib/domain/services/local_sync.service.dart
#	mobile/lib/platform/native_sync_api.g.dart
#	mobile/lib/providers/infrastructure/sync.provider.dart
#	mobile/pigeon/native_sync_api.dart
#	mobile/test/domain/services/hash_service_test.dart

Commit:
resolve merge conflicts
use updated approach for calculating checksums

Commit:
use CurrentPlatform instead _platform
fix mocks

Commit:
revert redundant changes

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
Include trashed items in getMediaChanges
Process trashed items delta during incremental sync

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/drift_schemas/main/drift_schema_v11.json
#	mobile/lib/infrastructure/repositories/db.repository.dart
#	mobile/lib/infrastructure/repositories/db.repository.steps.dart
#	mobile/test/drift/main/generated/schema_v11.dart

Commit:
fix merge conflicts

Commit:
fix format

Commit:
trashed_local_asset table mirror of local_asset table structure
trashed_local_asset<->local_asset transfer data on move to trash or restore
refactor code

Commit:
refactor and format code

Commit:
refactor TrashedAsset model
fix missed data transfering

Commit:
refactor code
remove unused model

Commit:
fix label

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/drift_schemas/main/drift_schema_v12.json
#	mobile/lib/infrastructure/repositories/db.repository.dart
#	mobile/lib/infrastructure/repositories/db.repository.steps.dart
#	mobile/test/drift/main/generated/schema.dart
#	mobile/test/drift/main/generated/schema_v12.dart

Commit:
fix merge conflicts

shenlong-tanwen reviewed (commented):
I see the restores are processed during the local sync. So maybe it makes sense to process the restores once after the remote sync is completed as well?

shenlong-tanwen commented:
> Maybe instead of moving it, we should add a call to `_applyRemoteRestoreToLocal ` whenever any `SyncAssetV1 ` is received?

I'm not actually suggest us moving the restore from the local sync process. My suggestion was more like having the`_applyRemoteRestoreToLocal` call in the when complete callback of the remote sync in `background_sync.dart` file

PeterOmbodi commented on mobile/lib/infrastructure/entities/trashed_local_asset.entity.dart:
```diff
@@ -0,0 +1,43 @@
+
+  TextColumn get id => text()();
+
+  TextColumn get albumId => text()();
```
this is possible, but as I understand it, a local asset has a unique ID (MediaStore.id within a volume, though we ignore volumes), and the albumId is tied to the hash of the file system path where it resides. This means that for a given asset ID, there can only be one albumId. In that case, an additional table would be redundant, as would the overhead of managing table relationships. This is also the reason for having a separate TrashedAsset model. Of course, I may not be considering all cases here, so please correct me if I’m wrong.

PeterOmbodi commented on mobile/lib/widgets/settings/beta_sync_settings/sync_status_and_actions.dart:
```diff
@@ -351,6 +353,42 @@ class _SyncStatsCounts extends ConsumerWidget {
               ),
             ),
+            if (trashSyncService.isAutoSyncMode) ...[
+              _SectionHeaderText(text: "trash".t(context: context)),
+              Consumer(
+                builder: (context, ref, _) {
+                  final counts = ref.watch(trashedAssetsCountProvider);
+                  return counts.when(
```
This part is only shown when the experimental trash sync feature flag is enabled. It’s convenient here to trigger sync and hashing and observe the results and processing speed. If this is unnecessary, I can remove it. Or we could remove it after testing.

PeterOmbodi commented on mobile/lib/infrastructure/repositories/local_asset.repository.dart:
```diff
@@ -95,4 +99,31 @@ class DriftLocalAssetRepository extends DriftDatabaseRepository {
     return query.map((localAlbum) => localAlbum.toDto()).get();
   }
+
+  Future<Map<AlbumId, List<LocalAsset>>> getBackupSelectedAssetsByAlbum(Iterable<String> checksums) async {
```
I meant that the assets are grouped by albums. But you’re right, the method should be renamed.

PeterOmbodi commented on mobile/lib/domain/models/asset/trashed_asset.model.dart:
```diff
@@ -0,0 +1,124 @@
+import 'package:immich_mobile/domain/models/asset/base_asset.model.dart';
+
+class TrashedAsset extends LocalAsset {
```
see https://github.com/immich-app/immich/pull/20473#discussion_r2399048480

PeterOmbodi committed:
optimize, refactor code
remove redundant code and checking
getTrashedAssetsForAlbum for iOS
tests for hash trashed assets

PeterOmbodi committed:
format code

PeterOmbodi committed:
fix migration
fix tests

PeterOmbodi committed:
fix generated file

PeterOmbodi committed:
reuse exist checksums on trash data update
handle restoration errors
fix import

PeterOmbodi committed:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

PeterOmbodi committed:
format code

PeterOmbodi committed:
sync_stream.service depend on repos
refactor assets restoration
update dependencies in tests

PeterOmbodi committed:
remove trashed asset model
remove trash_sync.service
refactor DriftTrashedLocalAssetRepository, LocalSyncService

PeterOmbodi committed:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

shenlong-tanwen commented on mobile/pigeon/native_sync_api.dart:
```diff
@@ -107,4 +123,11 @@ abstract class NativeSyncApi {
   void cancelHashing();
+
+  @TaskQueue(type: TaskQueueType.serialBackgroundThread)
+  List<PlatformAsset> getTrashedAssetsForAlbum(String albumId);
```
Maybe add a new API called getTrashChanges that sends the delta for the trash instead of a param on the getMediaChanges call? Also, the delta sync and full sync should not have any implementation difference. If we are syncing everything on the delta sync, we should also be doing the same on the full sync as well.

Commit:
rework fetching trashed assets data on native side
optimize handling trashed assets in local sync service
refactor code

PeterOmbodi committed:
update NativeSyncApi on iOS side
remove unused code

PeterOmbodi committed:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/ios/Podfile.lock

PeterOmbodi committed:
optimize sync trashed assets call in full sync mode
refactor code

PeterOmbodi committed:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

PeterOmbodi committed:
fix format

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
remove albumIds from getTrashedAssets params
fix upsert in trashed local asset repo
refactor code

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/ios/Runner/Sync/MessagesImpl.swift

Commit:
fix getTrashedAssets params

Commit:
fix(trash-sync): clean up NativeSyncApiImplBase and correct applyDelta

Commit:
refactor(trash-sync): optimize performance and fix minor issues

Commit:
refactor(trash-sync): add missed index

Commit:
feat(trash-sync): remove sinceLastCheckpoint param from getTrashedAssets

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/ios/Runner/Sync/MessagesImpl.swift

Commit:
fix(trash-sync): fix target table

Commit:
fix(trash-sync): remove unused extension

Commit:
fix(trash-sync): remove unused code

shenlong-tanwen commented on mobile/lib/infrastructure/repositories/trashed_local_asset.repository.dart:
```diff
@@ -0,0 +1,226 @@
+  /// Applies resulted snapshot of trashed assets:
+  /// - upserts incoming rows
+  /// - deletes rows that are not present in the snapshot
+  Future<void> applyTrashedAssets(Iterable<TrashedAsset> trashedAssets) async {
```
Maybe rename this to `processSnapshot` or something?

Commit:
fix(trash-sync): refactor code

PeterOmbodi committed:
Merge branch 'immich-app:main' into feature/sync_assets_trashed_state

Commit:
Merge remote-tracking branch 'origin/feature/sync_assets_trashed_state' into feature/sync_assets_trashed_state

Commit:
fix(trash-sync): reformat file

PeterOmbodi commented on mobile/lib/infrastructure/repositories/trashed_local_asset.repository.dart:
```diff
@@ -0,0 +1,226 @@
+  final Drift _db;
+
+  const DriftTrashedLocalAssetRepository(this._db) : super(_db);
+  static const _chunk = 32000;
```
added into constants
`const int kDriftMaxChunk = 32000;`

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/lib/constants/constants.dart

Commit:
fix(trash_sync): refactor code

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
fix(trash_sync): improve moving to trash

Commit:
refactor(trash_sync): integrate MANAGE_MEDIA permission request into login flow and advanced settings

Commit:
refactor(trash_sync): add additional checking for experimental trash sync flag and MANAGE_MEDIA permission.

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/lib/platform/native_sync_api.g.dart

Commit:
refactor(trash_sync): resolve merge conflicts

Commit:
refactor(trash_sync): fix format

PeterOmbodi committed:
Merge branch 'main' into feature/sync_assets_trashed_state

PeterOmbodi committed:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/lib/widgets/forms/login/login_form.dart
#	mobile/test/infrastructure/repository.mock.dart

PeterOmbodi committed:
resolve merge conflicts
add await for alert dialog
add missed request

Commit:
refactor(trash_sync): rework MANAGE_MEDIA info widget
show rationale text in permission request alert dialog
refactor setting getter

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
fix(trash_sync): restore missing text values

PeterOmbodi committed:
fix(trash_sync): format file

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

# Conflicts:
#	mobile/lib/constants/constants.dart

Commit:
Merge remote-tracking branch 'origin/main' into feature/sync_assets_trashed_state

alextran1502 committed:
Merge branch 'main' into feature/sync_assets_trashed_state

Commit:
fix(trash_sync): check backup enabled and remove remote asset existence check

Commit:
fix(trash_sync): remove checking backup enabled
test(trash_sync): cover sync-stream trash/restore paths and dedupe mocks

Commit:
test(trash_sync): cover trash/restore flows for local_sync_service

Commit:
chore(e2e): restore test-assets submodule pointer

alextran1502 committed:
Merge branch 'main' into feature/sync_assets_trashed_state
